### PR TITLE
[ETP-546] Abstract email sending process for tickets

### PR DESCRIPTION
**Pull request by sc0ttkclark**
[ETP-546]

Dev screencast: https://share.skc.dev/o0umlBAK

Related to the Event Tickets Plus PR: https://github.com/moderntribe/event-tickets-plus/pull/1084

This PR abstracts email sending process for Tribe Commerce.

[ETP-546]: https://moderntribe.atlassian.net/browse/ETP-546

### CODE REVIEW

File: src/Tribe/Commerce/PayPal/Main.php [hunk 1]
```diff
@@ -1099,100 +1098,49 @@ public function send_tickets_email( $order_id, $post_id = null ) {
...
-		 * @param string $order_id
-		 */
-		$subject = apply_filters( 'tribe_tpp_email_subject', tribe_get_option( 'ticket-paypal-confirmation-email-subject', $default_subject ) );
+	/**
+	 * Send RSVPs/tickets email for attendees.
+	 *
+	 * @since TBD
+	 *
+	 * @param array $attendees List of attendees.
+	 * @param array $args      {
```
Comment (tr1b0t) [inline src/Tribe/Commerce/PayPal/Main.php:1116]: :no_entry_sign: Parameter comment must end with a full stop

<sub>hash: 01b91d1397f2d117255fc78ca2ff7bd4</sub>

File: src/Tribe/Tickets.php [hunk 2]
```diff
@@ -2110,6 +2110,324 @@ public function generate_tickets_email_content( $tickets ) {
 			return tribe_tickets_get_template_part( 'tickets/email', null, [ 'tickets' => $tickets ], false );
 		}
 
+		/**
+		 * Send RSVPs/tickets email for attendees.
+		 *
+		 * @since TBD
+		 *
+		 * @param array $attendees List of attendees.
+		 * @param array $args      {
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:2119]: :no_entry_sign: Parameter comment must end with a full stop

<sub>hash: d91030f62312f5a840cb0d631297670b</sub>

File: src/Tribe/Tickets.php [hunk 3]
```diff
@@ -2110,6 +2110,324 @@ public function generate_tickets_email_content( $tickets ) {
...
+		}
+
+		/**
+		 * Send RSVPs/tickets email for an attendee.
+		 *
+		 * @since TBD
+		 *
+		 * @param string $to      The email to send the tickets to.
+		 * @param array  $tickets The list of tickets to send.
+		 * @param array  $args    {
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:2163]: :no_entry_sign: Parameter comment must end with a full stop

<sub>hash: 69b7ef72830b3b6f9976f28fe57b76b1</sub>

File: src/Tribe/Tickets.php [hunk 4]
```diff
@@ -2110,6 +2110,324 @@ public function generate_tickets_email_content( $tickets ) {
...
+				$singular = $is_rsvp
+					? tribe_get_rsvp_label_singular( 'RSVP email send' )
+					: tribe_get_ticket_label_singular_lowercase( 'ticket email send' );
+
+				$plural = $is_rsvp
+					? tribe_get_rsvp_label_plural( 'RSVPs email send' )
+					: tribe_get_ticket_label_plural_lowercase( 'tickets email send' );
+
+				// translators: %1$s: The singular of "RSVP" or "ticket", %2$s: The plural of "RSVPs" or "tickets", %3$s: The site name.
+				$subject_string = _nx( 'Your %1$s from %3$s', 'Your %2$s from %3$s', count( $tickets ), 'The default RSVP/ticket email subject', 'event-tickets' );
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:2237]: :warning: Mismatched placeholders is probably an error

<sub>hash: 7e5521712b55af260da734073d84e007</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 3 error(s)
:warning: 1 warning(s)

File: src/Tribe/Tickets.php [hunk 5]
```diff
@@ -2110,6 +2110,324 @@ public function generate_tickets_email_content( $tickets ) {
...
+			 * The dynamic portion of the filter hook, `$provider`, refers to the provider slug (rsvp, tpp, woo, edd).
+			 *
+			 * @since TBD
+			 *
+			 * @param string     $to       The email to send to.
+			 * @param int        $post_id  The post/event ID to send the email for.
+			 * @param string|int $order_id The order ID to send the email for.
+			 * @param array      $tickets  The list of tickets to send.
+			 */
+			$to = apply_filters( "tribe_{$provider}_email_recipient", $to, $post_id, $order_id, $tickets );
```
Comment (juanfra) [inline src/Tribe/Tickets.php:2285]: Maybe add the `_tickets` prefix before provider to keep in line with things?

File: src/Tribe/Tickets.php [hunk 6]
```diff
@@ -2110,6 +2110,324 @@ public function generate_tickets_email_content( $tickets ) {
...
+			 * The dynamic portion of the filter hook, `$provider`, refers to the provider slug (rsvp, tpp, woo, edd).
+			 *
+			 * @since TBD
+			 *
+			 * @param string     $subject  The email subject.
+			 * @param int        $post_id  The post/event ID to send the email for.
+			 * @param string|int $order_id The order ID to send the email for.
+			 * @param array      $tickets  The list of tickets to send.
+			 */
+			$subject = apply_filters( "tribe_{$provider}_email_subject", $subject, $post_id, $order_id, $tickets );
```
Comment (juanfra) [inline src/Tribe/Tickets.php:2316]: Maybe add the `_tickets` prefix before provider to keep in line with things?

File: src/Tribe/Tickets.php [hunk 7]
```diff
@@ -2110,6 +2110,324 @@ public function generate_tickets_email_content( $tickets ) {
...
+			 * The dynamic portion of the filter hook, `$provider`, refers to the provider slug (rsvp, tpp, woo, edd).
+			 *
+			 * @since TBD
+			 *
+			 * @param array      $content  The content to send the email with.
+			 * @param int        $post_id  The post/event ID to send the email for.
+			 * @param string|int $order_id The order ID to send the email for.
+			 * @param array      $tickets  The list of tickets to send.
+			 */
+			$content = apply_filters( "tribe_{$provider}_email_content", $this->generate_tickets_email_content( $tickets ), $post_id, $order_id, $tickets );
```
Comment (juanfra) [inline src/Tribe/Tickets.php:2347]: Maybe add the `_tickets` prefix before provider to keep in line with things?

File: src/Tribe/Tickets.php [hunk 8]
```diff
@@ -2110,6 +2110,324 @@ public function generate_tickets_email_content( $tickets ) {
...
+			 * The dynamic portion of the filter hook, `$provider`, refers to the provider slug (rsvp, tpp, woo, edd).
+			 *
+			 * @since TBD
+			 *
+			 * @param array      $headers  List of email headers.
+			 * @param int        $post_id  The post/event ID to send the email for.
+			 * @param string|int $order_id The order ID to send the email for.
+			 * @param array      $tickets  The list of tickets to send.
+			 */
+			$headers = apply_filters( "tribe_{$provider}_email_headers", $headers, $post_id, $order_id, $tickets );
```
Comment (juanfra) [inline src/Tribe/Tickets.php:2378]: Maybe add the `_tickets` prefix before provider to keep in line with things?

File: src/Tribe/Tickets.php [hunk 9]
```diff
@@ -2110,6 +2110,324 @@ public function generate_tickets_email_content( $tickets ) {
...
+			 * The dynamic portion of the filter hook, `$provider`, refers to the provider slug (rsvp, tpp, woo, edd).
+			 *
+			 * @since TBD
+			 *
+			 * @param array      $attachments The list of attachments to send.
+			 * @param int        $post_id     The post/event ID to send the email for.
+			 * @param string|int $order_id    The order ID to send the email for.
+			 * @param array      $tickets     The list of tickets to send.
+			 */
+			$attachments = apply_filters( "tribe_{$provider}_email_attachments", $attachments, $post_id, $order_id, $tickets );
```
Comment (juanfra) [inline src/Tribe/Tickets.php:2404]: Maybe add the `_tickets` prefix before provider to keep in line with things?

**Review by juanfra — APPROVED** [review]
Looking great! just a couple comments about the filter names.

File: src/Tribe/Tickets.php (hunk 5 above)
Comment (sc0ttkclark) [reply src/Tribe/Tickets.php:2285]: This is just backcompat for the old filter names :( not much I could do there unfortunately, but we can remove them once we start recommending the better ones.

File: src/Tribe/Tickets.php [hunk 10]
```diff
@@ -2110,6 +2110,353 @@ public function generate_tickets_email_content( $tickets ) {
...
+				$singular = $is_rsvp
+					? tribe_get_rsvp_label_singular( 'RSVP email send' )
+					: tribe_get_ticket_label_singular_lowercase( 'ticket email send' );
+
+				$plural = $is_rsvp
+					? tribe_get_rsvp_label_plural( 'RSVPs email send' )
+					: tribe_get_ticket_label_plural_lowercase( 'tickets email send' );
+
+				// translators: %1$s: The singular of "RSVP" or "ticket", %2$s: The plural of "RSVPs" or "tickets", %3$s: The site name.
+				$subject_string = _nx( 'Your %1$s from %3$s', 'Your %2$s from %3$s', count( $tickets ), 'The default RSVP/ticket email subject', 'event-tickets' );
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:2241]: :warning: Mismatched placeholders is probably an error

<sub>hash: 0ad9bc89d638c3b83cffd635418846cf</sub>

**Review by tr1b0t — COMMENTED** [review]
:warning: 1 warning(s)

**Review by juanfra — APPROVED** [review]

**Pull request merged**